// @taquito/taquito/dist/lib/signer/noop.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopSigner = exports.UnconfiguredSignerError = void 0;
/**
 *  @category Error
 *  @description Error that indicates the signer has been unconfigured in the TezosToolkit instance
 */
class UnconfiguredSignerError extends Error {
    constructor() {
        super('No signer has been configured. Please configure one by calling setProvider({signer}) on your TezosToolkit instance.');
        this.name = 'UnconfiguredSignerError';
    }
}
exports.UnconfiguredSignerError = UnconfiguredSignerError;
/**
 * @description Default signer implementation which does nothing and produce invalid signature
 */
class NoopSigner {
    publicKey() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new UnconfiguredSignerError();
        });
    }
    publicKeyHash() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new UnconfiguredSignerError();
        });
    }
    secretKey() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new UnconfiguredSignerError();
        });
    }
    sign(_bytes, _watermark) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new UnconfiguredSignerError();
        });
    }
}
exports.NoopSigner = NoopSigner;
