// @taquito/taquito/dist/lib/operations/tx-rollup-batch-operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxRollupBatchOperation = void 0;
const operations_1 = require("./operations");
/**
 * @description TxRollupBatchOperation provides utility functions to fetch a newly issued operation of kind tx_rollup_submit_batch
 */
class TxRollupBatchOperation extends operations_1.Operation {
    constructor(hash, params, source, raw, results, context) {
        super(hash, raw, results, context);
        this.params = params;
        this.source = source;
    }
    get operationResults() {
        const txrollupBatchOp = Array.isArray(this.results) &&
            this.results.find((op) => op.kind === 'tx_rollup_submit_batch');
        const result = txrollupBatchOp && txrollupBatchOp.metadata && txrollupBatchOp.metadata.operation_result;
        return result ? result : undefined;
    }
    get status() {
        const operationResults = this.operationResults;
        if (operationResults) {
            return operationResults.status;
        }
        else {
            return 'unknown';
        }
    }
    get content() {
        return this.params.content;
    }
    get fee() {
        return this.params.fee;
    }
    get gasLimit() {
        return this.params.gas_limit;
    }
    get storageLimit() {
        return this.params.storage_limit;
    }
    get errors() {
        return this.operationResults && this.operationResults.errors;
    }
}
exports.TxRollupBatchOperation = TxRollupBatchOperation;
