// @taquito/taquito/dist/lib/operations/tx-rollup-origination-operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxRollupOriginationOperation = void 0;
const operations_1 = require("./operations");
/**
 * @description TxRollupOriginationOperation provides utility functions to fetch a newly issued operation of kind tx_rollup_origination
 */
class TxRollupOriginationOperation extends operations_1.Operation {
    constructor(hash, params, source, raw, results, context) {
        super(hash, raw, results, context);
        this.params = params;
        this.source = source;
        this.originatedRollup = this.operationResults && this.operationResults.originated_rollup;
    }
    get operationResults() {
        const rollupOriginationOp = Array.isArray(this.results) &&
            this.results.find((op) => op.kind === 'tx_rollup_origination');
        const result = rollupOriginationOp &&
            rollupOriginationOp.metadata &&
            rollupOriginationOp.metadata.operation_result;
        return result ? result : undefined;
    }
    get status() {
        const operationResults = this.operationResults;
        if (operationResults) {
            return operationResults.status;
        }
        else {
            return 'unknown';
        }
    }
    get fee() {
        return this.params.fee;
    }
    get gasLimit() {
        return this.params.gas_limit;
    }
    get storageLimit() {
        return this.params.storage_limit;
    }
    get errors() {
        return this.operationResults && this.operationResults.errors;
    }
}
exports.TxRollupOriginationOperation = TxRollupOriginationOperation;
