// @taquito/rpc/dist/lib/rpc-client-modules/rpc-cache.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RpcClientCache = void 0;
const rpc_client_interface_1 = require("./../rpc-client-interface");
const rpc_client_interface_2 = require("../rpc-client-interface");
const utils_1 = require("@taquito/utils");
const defaultTtl = 1000;
/***
 * @description RpcClientCache acts as a decorator over the RpcClient instance by caching responses for the period defined by the ttl.
 */
class RpcClientCache {
    /**
     *
     * @param rpcClient rpcClient responsible of the interaction with Tezos network through an rpc node
     * @param ttl number representing the time to live (default 1000 milliseconds)
     *
     * @example new RpcClientCache(new RpcClient('https://mainnet.api.tez.ie/'))
     */
    constructor(rpcClient, ttl = defaultTtl) {
        this.rpcClient = rpcClient;
        this.ttl = ttl;
        this._cache = {};
    }
    getAllCachedData() {
        return this._cache;
    }
    /**
     * @description Remove all the data in the cache.
     *
     */
    deleteAllCachedData() {
        for (const key in this._cache) {
            delete this._cache[key];
        }
    }
    formatCacheKey(rpcUrl, rpcMethodName, rpcMethodParams, rpcMethodData) {
        let paramsToString = '';
        rpcMethodParams.forEach((param) => {
            paramsToString =
                typeof param === 'object'
                    ? paramsToString + JSON.stringify(param) + '/'
                    : paramsToString + param + '/';
        });
        return rpcMethodData
            ? `${rpcUrl}/${rpcMethodName}/${paramsToString}/${JSON.stringify(rpcMethodData)}`
            : `${rpcUrl}/${rpcMethodName}/${paramsToString}`;
    }
    has(key) {
        return key in this._cache;
    }
    get(key) {
        return this._cache[key].response;
    }
    put(key, response) {
        const handle = setTimeout(() => {
            return this.remove(key);
        }, this.ttl);
        Object.assign(this._cache, { [key]: { handle, response } });
    }
    remove(key) {
        if (key in this._cache) {
            delete this._cache[key];
        }
    }
    validateAddress(address) {
        if (utils_1.validateAddress(address) !== utils_1.ValidationResult.VALID) {
            throw new utils_1.InvalidAddressError(address);
        }
    }
    validateContract(address) {
        if (utils_1.validateContractAddress(address) !== utils_1.ValidationResult.VALID) {
            throw new utils_1.InvalidContractAddressError(address);
        }
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Get the block's hash, its unique identifier.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-hash
     */
    getBlockHash({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BLOCK_HASH, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlockHash({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List the ancestors of the given block which, if referred to as the branch in an operation header, are recent enough for that operation to be included in the current block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-live-blocks
     */
    getLiveBlocks({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_LIVE_BLOCKS, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getLiveBlocks({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address address from which we want to retrieve the balance
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the balance of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-balance
     */
    getBalance(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BALANCE, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBalance(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the storage
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the data of the contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-storage
     */
    getStorage(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_STORAGE, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getStorage(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the script
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the code and data of the contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-script
     */
    getScript(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_SCRIPT, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getScript(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the script
     * @param unparsingMode default is { unparsing_mode: "Readable" }
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the script of the contract and normalize it using the requested unparsing mode.
     *
     */
    getNormalizedScript(address, unparsingMode = { unparsing_mode: 'Readable' }, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_NORMALIZED_SCRIPT, [block, address, unparsingMode]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getNormalizedScript(address, unparsingMode, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the complete status of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id
     */
    getContract(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_CONTRACT, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getContract(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the manager
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the manager key of a contract.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-manager-key
     */
    getManagerKey(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_MANAGER_KEY, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getManagerKey(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the delegate (baker)
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the delegate of a contract, if any.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-delegate
     */
    getDelegate(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_DELEGATE, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getDelegate(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address contract address from which we want to retrieve the big map key
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a key in the big map storage of the contract.
     *
     * @deprecated Deprecated in favor of getBigMapKeyByID
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-context-contracts-contract-id-big-map-get
     */
    getBigMapKey(address, key, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const keyUrl = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BIG_MAP_KEY, [
                block,
                address,
                key,
            ]);
            if (this.has(keyUrl)) {
                return this.get(keyUrl);
            }
            else {
                const response = this.rpcClient.getBigMapKey(address, key, { block });
                this.put(keyUrl, response);
                return response;
            }
        });
    }
    /**
     *
     * @param id Big Map ID
     * @param expr Expression hash to query (A b58check encoded Blake2b hash of the expression (The expression can be packed using the pack_data method))
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a key in a big map.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-big-maps-big-map-id-script-expr
     */
    getBigMapExpr(id, expr, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BIG_MAP_EXPR, [
                block,
                id,
                expr,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBigMapExpr(id, expr, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param address delegate address which we want to retrieve
     * @param options contains generic configuration for rpc calls
     *
     * @description Fetches information about a delegate from RPC.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-delegates-pkh
     */
    getDelegates(address, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateAddress(address);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_DELEGATES, [
                block,
                address,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getDelegates(address, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description All constants
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-constants
     */
    getConstants({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_CONSTANTS, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getConstants({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls. See examples for various available sytaxes.
     *
     * @description All the information about a block
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id
     * @example getBlock() will default to /main/chains/block/head.
     * @example getBlock({ block: head~2 }) will return an offset of 2 blocks.
     * @example getBlock({ block: BL8fTiWcSxWCjiMVnDkbh6EuhqVPZzgWheJ2dqwrxYRm9AephXh~2 }) will return an offset of 2 blocks from given block hash..
     */
    getBlock({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BLOCK, [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlock({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description The whole block header
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-header
     */
    getBlockHeader({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BLOCK_HEADER, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlockHeader({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description All the metadata associated to the block
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-metadata
     */
    getBlockMetadata({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BLOCK_METADATA, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBlockMetadata({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param args contains optional query arguments
     * @param options contains generic configuration for rpc calls
     *
     * @description Retrieves the list of delegates allowed to bake a block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-helpers-baking-rights
     */
    getBakingRights(args = {}, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BAKING_RIGHTS, [
                block,
                args,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBakingRights(args, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param args contains optional query arguments
     * @param options contains generic configuration for rpc calls
     *
     * @description Retrieves the list of delegates allowed to bake a block.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-helpers-endorsing-rights
     */
    getEndorsingRights(args = {}, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_ENDORSING_RIGHTS, [block, args]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getEndorsingRights(args, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     * @param options contains generic configuration for rpc calls
     *
     * @description Ballots casted so far during a voting period
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-ballot-list
     */
    getBallotList({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BALLOT_LIST, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBallotList({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Sum of ballots casted so far during a voting period.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-ballots
     */
    getBallots({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_BALLOTS, [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getBallots({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current proposal under evaluation.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-proposal
     */
    getCurrentProposal({ block, } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_CURRENT_PROPOSAL, [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentProposal({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Current expected quorum.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-quorum
     */
    getCurrentQuorum({ block, } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_CURRENT_QUORUM, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentQuorum({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List of delegates with their voting weight, in number of rolls.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-listings
     */
    getVotesListings({ block, } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_VOTES_LISTINGS, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getVotesListings({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description List of proposals with number of supporters.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-proposals
     */
    getProposals({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_PROPOSALS, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getProposals({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param data operation contents to forge
     * @param options contains generic configuration for rpc calls
     *
     * @description Forge an operation returning the unsigned bytes
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-forge-operations
     */
    forgeOperations(data, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.forgeOperations(data, { block });
        });
    }
    /**
     *
     * @param signedOpBytes signed bytes to inject
     *
     * @description Inject an operation in node and broadcast it. Returns the ID of the operation. The `signedOperationContents` should be constructed using a contextual RPCs from the latest block and signed by the client. By default, the RPC will wait for the operation to be (pre-)validated before answering. See RPCs under /blocks/prevalidation for more details on the prevalidation context.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-injection-operation
     */
    injectOperation(signedOpBytes) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.injectOperation(signedOpBytes);
        });
    }
    /**
     *
     * @param ops Operations to apply
     * @param options contains generic configuration for rpc calls
     *
     * @description Simulate the validation of an operation
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-preapply-operations
     */
    preapplyOperations(ops, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.preapplyOperations(ops, { block });
        });
    }
    /**
     *
     * @param contract address of the contract we want to get the entrypoints of
     *
     * @description Return the list of entrypoints of the contract
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-context-contracts-contract-id-entrypoints
     *
     * @version 005_PsBABY5H
     */
    getEntrypoints(contract, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            this.validateContract(contract);
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_ENTRYPOINTS, [
                block,
                contract,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getEntrypoints(contract, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     * @param op Operation to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Run an operation without signature checks
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-run-operation
     */
    runOperation(op, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.runOperation(op, { block });
        });
    }
    /**
     * @param code Code to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Run a piece of code in the current context
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-run-code
     */
    runCode(code, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.runCode(code, { block });
        });
    }
    /**
     * @param viewParams Parameters of the view to run
     * @param options contains generic configuration for rpc calls
     *
     * @description Simulate a call to a view following the TZIP-4 standard. See https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-4/tzip-4.md#view-entrypoints.
     *
     */
    runView(_a, { block } = rpc_client_interface_2.defaultRPCOptions) {
        var { unparsing_mode = 'Readable' } = _a, rest = __rest(_a, ["unparsing_mode"]);
        return __awaiter(this, void 0, void 0, function* () {
            return this.rpcClient.runView(Object.assign({ unparsing_mode }, rest), { block });
        });
    }
    getChainId() {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_CHAIN_ID, []);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getChainId();
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param data Data to pack
     * @param options contains generic configuration for rpc calls
     *
     * @description Computes the serialized version of a data expression using the same algorithm as script instruction PACK
     *
     * @example packData({ data: { string: "test" }, type: { prim: "string" } })
     *
     * @see https://tezos.gitlab.io/api/rpc.html#post-block-id-helpers-scripts-pack-data
     */
    packData(data, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.PACK_DATA, [
                block,
                data,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.packData(data, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @description Return rpc root url
     */
    getRpcUrl() {
        return this.rpcClient.getRpcUrl();
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Voting period of current block.
     *
     * @example getCurrentPeriod() will default to current voting period for /main/chains/block/head.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-current-period
     */
    getCurrentPeriod({ block, } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_CURRENT_PERIOD, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getCurrentPeriod({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param options contains generic configuration for rpc calls
     *
     * @description Voting period of next block.
     *
     * @example getSuccessorPeriod() will default to successor voting period for /main/chains/block/head.
     *
     * @see https://tezos.gitlab.io/api/rpc.html#get-block-id-votes-successor-period
     */
    getSuccessorPeriod({ block, } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_SUCCESSOR_PERIOD, [block]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getSuccessorPeriod({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param id Sapling state ID
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a sapling state ID.
     *
     * @see https://tezos.gitlab.io/active/rpc.html#get-block-id-context-sapling-sapling-state-id-get-diff
     */
    getSaplingDiffById(id, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_SAPLING_DIFF_BY_ID, [block, id]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getSaplingDiffById(id, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    /**
     *
     * @param contract address of the contract we want to get the sapling diff
     * @param options contains generic configuration for rpc calls
     *
     * @description Access the value associated with a sapling state.
     *
     * @see https://tezos.gitlab.io/active/rpc.html#get-block-id-context-contracts-contract-id-single-sapling-get-diff
     */
    getSaplingDiffByContract(contract, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_SAPLING_DIFF_BY_CONTRACT, [block, contract]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getSaplingDiffByContract(contract, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    getProtocols({ block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_PROTOCOLS, [
                block,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getProtocols({ block });
                this.put(key, response);
                return response;
            }
        });
    }
    getTxRollupState(txRollupId, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_TX_ROLLUP_STATE, [
                block,
                txRollupId,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getTxRollupState(txRollupId, { block });
                this.put(key, response);
                return response;
            }
        });
    }
    getTxRollupInbox(txRollupId, blockLevel, { block } = rpc_client_interface_2.defaultRPCOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            const key = this.formatCacheKey(this.rpcClient.getRpcUrl(), rpc_client_interface_1.RPCMethodName.GET_TX_ROLLUP_INBOX, [
                block,
                txRollupId,
                blockLevel,
            ]);
            if (this.has(key)) {
                return this.get(key);
            }
            else {
                const response = this.rpcClient.getTxRollupInbox(txRollupId, blockLevel, { block });
                this.put(key, response);
                return response;
            }
        });
    }
}
exports.RpcClientCache = RpcClientCache;
