// @taquito/michelson-encoder/dist/lib/tokens/comparable/bytes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BytesToken = exports.BytesValidationError = void 0;
const token_1 = require("../token");
class BytesValidationError extends token_1.TokenValidationError {
    constructor(value, token, message) {
        super(value, token, message);
        this.value = value;
        this.token = token;
        this.name = 'BytesValidationError';
    }
}
exports.BytesValidationError = BytesValidationError;
class BytesToken extends token_1.ComparableToken {
    constructor(val, idx, fac) {
        super(val, idx, fac);
        this.val = val;
        this.idx = idx;
        this.fac = fac;
    }
    ToBigMapKey(val) {
        return {
            key: { bytes: val },
            type: { prim: BytesToken.prim },
        };
    }
    isValid(val) {
        if (typeof val === 'string' && /^[0-9a-fA-F]*$/.test(val) && val.length % 2 === 0) {
            return null;
        }
        else {
            return new BytesValidationError(val, this, `Invalid bytes: ${val}`);
        }
    }
    convertUint8ArrayToHexString(val) {
        return val.constructor === Uint8Array ? Buffer.from(val).toString('hex') : val;
    }
    Encode(args) {
        let val = args.pop();
        val = this.convertUint8ArrayToHexString(val);
        const err = this.isValid(val);
        if (err) {
            throw err;
        }
        return { bytes: String(val).toString() };
    }
    EncodeObject(val, semantic) {
        val = this.convertUint8ArrayToHexString(val);
        const err = this.isValid(val);
        if (err) {
            throw err;
        }
        if (semantic && semantic[BytesToken.prim]) {
            return semantic[BytesToken.prim](val);
        }
        return { bytes: String(val).toString() };
    }
    Execute(val) {
        return val.bytes;
    }
    /**
     * @deprecated ExtractSchema has been deprecated in favor of generateSchema
     *
     */
    ExtractSchema() {
        return BytesToken.prim;
    }
    generateSchema() {
        return {
            __michelsonType: BytesToken.prim,
            schema: BytesToken.prim,
        };
    }
    ToKey({ bytes, string }) {
        if (string) {
            return string;
        }
        return bytes;
    }
    findAndReturnTokens(tokenToFind, tokens) {
        if (BytesToken.prim === tokenToFind) {
            tokens.push(this);
        }
        return tokens;
    }
}
exports.BytesToken = BytesToken;
BytesToken.prim = 'bytes';
