// @taquito/michelson-encoder/dist/lib/tokens/bls12-381-fr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Bls12381frToken = exports.Bls12381frValidationError = void 0;
const token_1 = require("./token");
class Bls12381frValidationError extends token_1.TokenValidationError {
    constructor(value, token, message) {
        super(value, token, message);
        this.value = value;
        this.token = token;
        this.name = 'Bls12381frValidationError';
    }
}
exports.Bls12381frValidationError = Bls12381frValidationError;
class Bls12381frToken extends token_1.Token {
    constructor(val, idx, fac) {
        super(val, idx, fac);
        this.val = val;
        this.idx = idx;
        this.fac = fac;
    }
    isValid(val) {
        if (/^[0-9a-fA-F]*$/.test(val) && val.length % 2 === 0) {
            return null;
        }
        else {
            return new Bls12381frValidationError(val, this, `Invalid bytes: ${val}`);
        }
    }
    convertUint8ArrayToHexString(val) {
        return val.constructor === Uint8Array ? Buffer.from(val).toString('hex') : val;
    }
    Encode(args) {
        let val = args.pop();
        if (typeof val === 'number') {
            return { int: val.toString() };
        }
        else {
            val = this.convertUint8ArrayToHexString(val);
            const err = this.isValid(val);
            if (err) {
                throw err;
            }
            return { bytes: val };
        }
    }
    EncodeObject(val, semantic) {
        if (semantic && semantic[Bls12381frToken.prim]) {
            return semantic[Bls12381frToken.prim](val);
        }
        if (typeof val === 'number') {
            return { int: val.toString() };
        }
        else {
            val = this.convertUint8ArrayToHexString(val);
            const err = this.isValid(val);
            if (err) {
                throw err;
            }
            return { bytes: val };
        }
    }
    Execute(val) {
        return val.bytes;
    }
    /**
     * @deprecated ExtractSchema has been deprecated in favor of generateSchema
     *
     */
    ExtractSchema() {
        return Bls12381frToken.prim;
    }
    generateSchema() {
        return {
            __michelsonType: Bls12381frToken.prim,
            schema: Bls12381frToken.prim,
        };
    }
    findAndReturnTokens(tokenToFind, tokens) {
        if (Bls12381frToken.prim === tokenToFind) {
            tokens.push(this);
        }
        return tokens;
    }
}
exports.Bls12381frToken = Bls12381frToken;
// An element of the BLS12-381 scalar field Fr
// see https://tezos.gitlab.io/michelson-reference/#type-bls12_381_fr
Bls12381frToken.prim = 'bls12_381_fr';
