// @taquito/rpc/dist/lib/opkind.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpKind = void 0;
var OpKind;
(function (OpKind) {
    OpKind["ORIGINATION"] = "origination";
    OpKind["DELEGATION"] = "delegation";
    OpKind["REVEAL"] = "reveal";
    OpKind["TRANSACTION"] = "transaction";
    OpKind["ACTIVATION"] = "activate_account";
    OpKind["ENDORSEMENT"] = "endorsement";
    OpKind["PREENDORSEMENT"] = "preendorsement";
    OpKind["SET_DEPOSITS_LIMIT"] = "set_deposits_limit";
    OpKind["DOUBLE_PREENDORSEMENT_EVIDENCE"] = "double_preendorsement_evidence";
    OpKind["ENDORSEMENT_WITH_SLOT"] = "endorsement_with_slot";
    OpKind["SEED_NONCE_REVELATION"] = "seed_nonce_revelation";
    OpKind["DOUBLE_ENDORSEMENT_EVIDENCE"] = "double_endorsement_evidence";
    OpKind["DOUBLE_BAKING_EVIDENCE"] = "double_baking_evidence";
    OpKind["PROPOSALS"] = "proposals";
    OpKind["BALLOT"] = "ballot";
    OpKind["FAILING_NOOP"] = "failing_noop";
    OpKind["REGISTER_GLOBAL_CONSTANT"] = "register_global_constant";
    OpKind["TX_ROLLUP_ORIGINATION"] = "tx_rollup_origination";
    OpKind["TX_ROLLUP_SUBMIT_BATCH"] = "tx_rollup_submit_batch";
    OpKind["TX_ROLLUP_COMMIT"] = "tx_rollup_commit";
    OpKind["TX_ROLLUP_RETURN_BOND"] = "tx_rollup_return_bond";
    OpKind["TX_ROLLUP_FINALIZE_COMMITMENT"] = "tx_rollup_finalize_commitment";
    OpKind["TX_ROLLUP_REMOVE_COMMITMENT"] = "tx_rollup_remove_commitment";
    OpKind["TX_ROLLUP_REJECTION"] = "tx_rollup_rejection";
    OpKind["TX_ROLLUP_DISPATCH_TICKETS"] = "tx_rollup_dispatch_tickets";
    OpKind["TRANSFER_TICKET"] = "transfer_ticket";
})(OpKind = exports.OpKind || (exports.OpKind = {}));
